app_flutter/lib/agent_list.dart

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import 'package:cocoon_service/protos.dart' show Agent;
+
+import 'agent_health_details.dart';
+import 'agent_tile.dart';
+import 'state/agent.dart';
+
+class AgentList extends StatefulWidget {
+  const AgentList(
+      {this.agents,
+      this.agentState,
+      this.agentFilter,
+      @visibleForTesting this.insertKeys = false});
+
+  final List<Agent> agents;
+
+  final AgentState agentState;
+
+  /// Term to filter [agents] by.
+  final String agentFilter;
+
+  /// When true, will set a key for the [AgentTile] that is composed of its
+  /// position in the list and the agent id.
+  @visibleForTesting
+  final bool insertKeys;
+
+  @override
+  _AgentListState createState() => _AgentListState();
+}
+
+class _AgentListState extends State<AgentList> {
+  TextEditingController filterAgentsController = TextEditingController();
+
+  @override
+  void initState() {
+    super.initState();
+
+    /// When redirected from the build dashboard, display a specific agent.
+    if (widget.agentFilter != null && widget.agentFilter.isNotEmpty) {
+      filterAgentsController.text = widget.agentFilter;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.agents.isEmpty) {
+      return const CircularProgressIndicator();
+    }
+
+    final List<FullAgent> fullAgents = widget.agents
```

Comment:
Could you have the service filter agents instead of doing it in build?

Reply:
Do you think there's an advantage to this? I left it in the widget since I made this into more of a search functionality as CanvasKit doesn't work with Ctrl+F. I thought it would be easier to follow if it was in the widget that uses it.